### PR TITLE
Add toJSON methods

### PR DESCRIPTION
**Pull request by appellation**
**Please describe the changes this PR makes and why it should be merged:**
Finishing #1602 

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/structures/Guild.js [hunk 1]
```diff
@@ -1138,6 +1138,28 @@ class Guild extends Base {
     return this.name;
   }
 
+  toJSON() {
+    return super.toJSON([
+      'createdAt',
+      'createdTimestamp',
```
Comment (devsnek) [inline src/structures/Guild.js]: only do createdAt, not createdTimestamp (applies to all)

File: src/structures/Guild.js [hunk 2]
```diff
@@ -1138,6 +1138,28 @@ class Guild extends Base {
...
+      'joinedAt',
+      'iconURL',
+      'nameAcronym',
+      'splashURL',
+      'owner',
+      'afkChannel',
```
Comment (devsnek) [inline src/structures/Guild.js]: things like this should be the IDs not the full object e.g. `afkChannelID`

File: src/structures/GuildChannel.js [hunk 3]
```diff
@@ -421,6 +421,16 @@ class GuildChannel extends Channel {
   toString() {
     return `<#${this.id}>`;
   }
+
+  toJSON(extra = []) {
+    return super.toJSON([
+      'calculatedPosition',
+      'members',
```
Comment (devsnek) [inline src/structures/GuildChannel.js]: nothing with state should be in the JSON e.g. members for GuildChannel

File: src/structures/MessageMentions.js [hunk 4]
```diff
@@ -131,6 +132,10 @@ class MessageMentions {
     const id = data.id || data;
     return this.users.has(id) || this.channels.has(id) || this.roles.has(id);
   }
+
+  toJSON() {
+    return Util.flatten(this, ['members', 'channels']);
```
Comment (devsnek) [inline src/structures/MessageMentions.js]: things like arrays of objects should be flattened to their IDs

File: src/client/Client.js [hunk 5]
```diff
@@ -470,6 +470,18 @@ class Client extends EventEmitter {
...
 
+  toJSON() {
+    return Util.flatten(this, [
+      'status',
+      'uptime',
+      'ping',
+      'voiceConnections',
+      'emojis',
+      'readyTimestamp',
+      'browser',
```
Comment (devsnek) [inline src/client/Client.js]: don't put `browser` in

File: src/util/Util.js [hunk 6]
```diff
@@ -13,6 +13,43 @@ class Util {
...
+
+      const elemIsObj = isObject(element);
+      const value = elemIsObj && typeof element.valueOf === 'function' ? element.valueOf() : null;
+
+      // If it's a collection, make the array of keys
+      if (element instanceof require('./Collection')) out[prop] = Array.from(element.keys());
+      // If it's an array, flatten each element
+      else if (Array.isArray(element)) out[prop] = element.map(e => Util.flatten(e));
+      // If it's a function, call it
+      else if (typeof element === 'function') out[prop] = element.call(obj);
```
Comment (devsnek) [inline src/util/Util.js]: this is an insanely bad idea. just ignore methods.

File: src/structures/MessageMentions.js (hunk 4 above)
Comment (appellation) [reply src/structures/MessageMentions.js]: Collections are automatically flattened to an array of their keys.

**Comment by appellation** [comment]
I have reduced the stateful properties as best I can, as requested.  There's a fine line between not snapshotting the object and providing a meaningful representation of it, so I hope I've found it; for instance, I removed all `edit` properties from the Message class, but left `content`, `attachments`, etc. even though they could potentially change.

File: src/client/Client.js [hunk 7]
```diff
@@ -470,6 +470,10 @@ class Client extends EventEmitter {
     this._intervals.delete(interval);
   }
 
+  toJSON() {
+    return Util.flatten(this, ['emojis'], ['readyAt', 'broadcasts', 'pings', 'presences', 'domain']);
```
Comment (devsnek) [inline src/client/Client.js]: emojis comes from guilds, don't put it here

File: src/structures/Guild.js [hunk 8]
```diff
@@ -1138,6 +1138,22 @@ class Guild extends Base {
...
+    const json = super.toJSON([
+      'createdTimestamp',
+      'nameAcronym',
+    ], [
+      'presences',
+      'voiceStates',
+      'verificationLevel',
+      'explicitContentFilter',
+      'features',
+    ]);
```
Comment (devsnek) [inline src/structures/Guild.js]: verification level, ecf, and features are all static information. For all of these methods, think about what the http api will give you, and ignore anything the ws gateway will give you.

File: src/util/Util.js [hunk 9]
```diff
@@ -13,6 +13,43 @@ class Util {
   }
 
   /**
+   * Flatten an object. Any properties that are collections will get converted to an array of keys.
+   * @param {Object} obj The object to flatten.
+   * @param {Array<Object<string, boolean|string>>} [props=[]] Specific properties to include/exclude.
```
Comment (bdistin) [inline src/util/Util.js]: After discussion about this param in staff chat, it should be: `...Object<string, boolean|string>`

**Pull request merged**